File: Java/P3.java
public class P3 {
    public static void main(String[] args){
        long number = 600851475143L;
        long p = 1;

        while(p < number){
            p = smallest_factor(number);
            number /= p;
        }

        System.out.println(p);
    }

    //Smallest factor will always be prime
    public static long smallest_factor(long n){
        int limit = (int) Math.ceil(Math.sqrt(n));
        for(int i = 3; i < limit; i+=2){
            if(n % i == 0) {
                return i;
            }
        }
        return n; // n itself is prime
    }
}

File: Rust/P3.rs
fn main(){
    let mut number: i64= 600851475143;
    let mut p: i64 = 1;

    while p < number {
        p = smallest_factor(number);
        number /= p;
    }

    println!("{}", p);
}

// The smallest factor will always be prime

fn smallest_factor(n: i64) -> i64{
    let limit = (n as f64).sqrt().ceil() as i64;
    
    for x in (3..limit).step_by(2) {
        if n % x == 0  {
            return x;
        }
    }
    n //n itself is prime
}
